[Remote Inspection] Add a way to immediately apply visibility adjustment based on CSS selectors when loading

whsieh opened the pull request:
#### 60ad4ce2c9dba49e168cb8048022d1e9baf67afc
<pre>
[Remote Inspection] Add a way to immediately apply visibility adjustment based on CSS selectors when loading
<a href="https://bugs.webkit.org/show_bug.cgi?id=272060">https://bugs.webkit.org/show_bug.cgi?id=272060</a>

Reviewed by Aditya Keerthi.

Add a mechanism which allows WebKit clients to eagerly apply visibility adjustment to elements per
navigation, based on sets of CSS selectors. See below for more details.

* Source/WebCore/loader/DocumentLoader.h:
(WebCore::DocumentLoader::visibilityAdjustmentSelectors const):
(WebCore::DocumentLoader::setVisibilityAdjustmentSelectors):

Add plumbing from `WebsitePoliciesData` down into `DocumentLoader`, like all of the other per-load
website policies.

* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::computeClientRect):
(WebCore::targetedElementInfo):
(WebCore::ElementTargetingController::adjustVisibilityInRepeatedlyTargetedRegions):
(WebCore::ElementTargetingController::applyVisibilityAdjustmentFromSelectors):

If we haven&apos;t already done so yet, pull `visibilityAdjustmentSelectors` from the `DocumentLoader`
into `ElementTargetingController`. For up to 30 seconds after committing the load, continually look
for elements that correspond to these selectors, and immediately adjust visibility for them if
found.

(WebCore::ElementTargetingController::reset):
(WebCore::ElementTargetingController::resetAdjustmentRegions): Deleted.

Rename `resetAdjustmentRegions` to just `reset`, since it now clears out more state than just what&apos;s
necessary for tracking adjustment regions.

* Source/WebCore/page/ElementTargetingController.h:
* Source/WebCore/page/ElementTargetingTypes.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::didCommitLoad):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebsitePoliciesData.cpp:
(WebKit::WebsitePoliciesData::applyToDocumentLoader):
* Source/WebKit/Shared/WebsitePoliciesData.h:
* Source/WebKit/Shared/WebsitePoliciesData.serialization.in:

Add more plumbing for CSS selectors from the client down into WebCore.

* Source/WebKit/UIProcess/API/APITargetedElementInfo.h:
* Source/WebKit/UIProcess/API/APIWebsitePolicies.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.mm:
(-[WKWebpagePreferences _setVisibilityAdjustmentSelectors:]):
(-[WKWebpagePreferences _visibilityAdjustmentSelectors]):

Add support for the new SPI. See above for more details.

* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferencesPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTargetedElementInfo.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTargetedElementInfo.mm:
(-[_WKTargetedElementInfo boundsInWebView]):
(-[_WKTargetedElementInfo boundsInClientCoordinates]):

Also, clean up this `bounds` SPI property — currently, the `-bounds` property returns the bounds of
the element in root view coordinates, while the targeted element request expects a point in the web
view&apos;s coordinate space. To address this inconsistency and clarify the behavior of the API, add
alternatives to `bounds`: `boundsInWebView` and `boundsInClientCoordinates` that explicitly indicate
which coordinate space the bounds are in. Depending on the purpose, the client can use either.

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ElementTargetingTests.mm:
(TestWebKitAPI::setUpWebViewForSnapshotting):

Factor this setup code into a common helper to reduce code duplication.

(TestWebKitAPI::TEST(ElementTargeting, AdjustVisibilityForUnparentedElement)):
(TestWebKitAPI::TEST(ElementTargeting, AdjustVisibilityFromSelectors)):
(TestWebKitAPI::TEST(ElementTargeting, AdjustVisibilityFromPseudoSelectors)):

Add an API test to exercise the new SPI, and verify that it works well with the rest of the existing
SPI surface around visibility adjustment.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/element-targeting-3.html: Added.
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:

(-[WKWebView synchronouslyLoadTestPageNamed:preferences:]):

Canonical link: <a href="https://commits.webkit.org/276990@main">https://commits.webkit.org/276990@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d018024830b8c48faec994bcb019a87cbc8d1268

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46358 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25498 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48951 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49033 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42399 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48665 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29854 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22954 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/37839 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46936 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/22544 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39966 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19070 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/19914 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41070 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4402 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/42644 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41430 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50860 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/21363 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/17815 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45064 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/22654 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43984 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10259 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/23023 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22353 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/page/ElementTargetingController.cpp:
```diff
@@ -702,13 +707,97 @@ void ElementTargetingController::adjustVisibilityInRepeatedlyTargetedRegions(Doc
+            if (pseudo == "after"_s)
+                return { components.first(), VisibilityAdjustment::AfterPseudo };
+
+            if (pseudo == "before"_s)
```
I think this comparison should be case-insensitive.

whsieh replied:
Great point — I'll account for this, and see if I can come up with a test case for it as well.

whsieh replied:
~(…will also take care of the `::after` case, as we chatted about in person)~

Just realized that this is going to match every element in the DOM... I think I'll do this in a separate patch, since it could require a bit more refactoring.

whsieh committed:
[Remote Inspection] Add a way to immediately apply visibility adjustment based on CSS selectors when loading
https://bugs.webkit.org/show_bug.cgi?id=272060

Reviewed by Aditya Keerthi.

Add a mechanism which allows WebKit clients to eagerly apply visibility adjustment to elements per
navigation, based on sets of CSS selectors. See below for more details.

* Source/WebCore/loader/DocumentLoader.h:
(WebCore::DocumentLoader::visibilityAdjustmentSelectors const):
(WebCore::DocumentLoader::setVisibilityAdjustmentSelectors):

Add plumbing from `WebsitePoliciesData` down into `DocumentLoader`, like all of the other per-load
website policies.

* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::computeClientRect):
(WebCore::targetedElementInfo):
(WebCore::ElementTargetingController::adjustVisibilityInRepeatedlyTargetedRegions):
(WebCore::ElementTargetingController::applyVisibilityAdjustmentFromSelectors):

If we haven't already done so yet, pull `visibilityAdjustmentSelectors` from the `DocumentLoader`
into `ElementTargetingController`. For up to 30 seconds after committing the load, continually look
for elements that correspond to these selectors, and immediately adjust visibility for them if
found.

(WebCore::ElementTargetingController::reset):
(WebCore::ElementTargetingController::resetAdjustmentRegions): Deleted.

Rename `resetAdjustmentRegions` to just `reset`, since it now clears out more state than just what's
necessary for tracking adjustment regions.

* Source/WebCore/page/ElementTargetingController.h:
* Source/WebCore/page/ElementTargetingTypes.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::didCommitLoad):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebsitePoliciesData.cpp:
(WebKit::WebsitePoliciesData::applyToDocumentLoader):
* Source/WebKit/Shared/WebsitePoliciesData.h:
* Source/WebKit/Shared/WebsitePoliciesData.serialization.in:

Add more plumbing for CSS selectors from the client down into WebCore.

* Source/WebKit/UIProcess/API/APITargetedElementInfo.h:
* Source/WebKit/UIProcess/API/APIWebsitePolicies.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.mm:
(-[WKWebpagePreferences _setVisibilityAdjustmentSelectors:]):
(-[WKWebpagePreferences _visibilityAdjustmentSelectors]):

Add support for the new SPI. See above for more details.

* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferencesPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTargetedElementInfo.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTargetedElementInfo.mm:
(-[_WKTargetedElementInfo boundsInWebView]):
(-[_WKTargetedElementInfo boundsInClientCoordinates]):

Also, clean up this `bounds` SPI property — currently, the `-bounds` property returns the bounds of
the element in root view coordinates, while the targeted element request expects a point in the web
view's coordinate space. To address this inconsistency and clarify the behavior of the API, add
alternatives to `bounds`: `boundsInWebView` and `boundsInClientCoordinates` that explicitly indicate
which coordinate space the bounds are in. Depending on the purpose, the client can use either.

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ElementTargetingTests.mm:
(TestWebKitAPI::setUpWebViewForSnapshotting):

Factor this setup code into a common helper to reduce code duplication.

(TestWebKitAPI::TEST(ElementTargeting, AdjustVisibilityForUnparentedElement)):
(TestWebKitAPI::TEST(ElementTargeting, AdjustVisibilityFromSelectors)):
(TestWebKitAPI::TEST(ElementTargeting, AdjustVisibilityFromPseudoSelectors)):

Add an API test to exercise the new SPI, and verify that it works well with the rest of the existing
SPI surface around visibility adjustment.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/element-targeting-3.html: Added.
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:

(-[WKWebView synchronouslyLoadTestPageNamed:preferences:]):

Canonical link: https://commits.webkit.org/276990@main
